### PR TITLE
FOUNDATIONS: Delete `Time`

### LINKED ISSUE #226: FOUNDATIONS: Delete `Time`

### PR DESCRIPTION
**Pull request by AbbosHaydarov**
Closes #226

### CODE REVIEW

File: Tarteeb.Api/Models/Foundations/Times/Exceptions/TimeValidationException.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+//=================================
+
+using Xeptions;
+
+namespace Tarteeb.Api.Models.Foundations.Times.Exceptions
+{
+    public class TimeValidationException : Xeption
+    {
+        public TimeValidationException(Xeption innerException)
+            : base(message: "Time validation error occured, fix the errors and try again.", innerException)
```
Comment (khabibulladev) [inline Tarteeb.Api/Models/Foundations/Times/Exceptions/TimeValidationException.cs:13]: occured -> occured

File: Tarteeb.Api/Services/Foundations/Times/TimeService.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            IDateTimeBroker dateTimeBroker, 
+            ILoggingBroker loggingBroker)
+        {
+            this.storageBroker = storageBroker;
+            this.dateTimeBroker = dateTimeBroker;
+            this.loggingBroker = loggingBroker;
+        }
+
+        public ValueTask<Time> RemoveTimeByIdAsync(Guid timeId) =>
+            TryCatch(async () =>
```
Comment (khabibulladev) [inline Tarteeb.Api/Services/Foundations/Times/TimeService.cs:32]: Remove a tab

**Review by khabibulladev — CHANGES_REQUESTED** [review]

**Comment by ElbekDeveloper** [comment]
startupda registratsiya qilinmagan

File: Tarteeb.Api/Models/Foundations/Times/Exceptions/TimeValidationException.cs (hunk 1 above)
Comment (AbbosHaydarov) [reply Tarteeb.Api/Models/Foundations/Times/Exceptions/TimeValidationException.cs:13]: hazillashdizooov aa Aylis 9+

File: Tarteeb.Api/Models/Foundations/Teams/Exceptions/NotFoundTeamException.cs [hunk 3]
```diff
@@ -11,7 +11,7 @@ namespace Tarteeb.Api.Models.Foundations.Teams.Exceptions
     public class NotFoundTeamException : Xeption
     {
         public NotFoundTeamException(Guid teamId)
-            : base(message: $"Couldn't find team with id: {teamId}.")
+            : base(message: $"Couldn't find time with id: {teamId}.")
```
Comment (GulchexrasGithub) [inline Tarteeb.Api/Models/Foundations/Teams/Exceptions/NotFoundTeamException.cs:14]: O'zi to'g'ri edi, time ni teamga o'zgartiring.

File: Tarteeb.Api/Services/Foundations/Times/TimeService.Validations.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+using Tarteeb.Api.Models.Foundations.Teams;
+using Tarteeb.Api.Models.Foundations.Times;
+using Tarteeb.Api.Models.Foundations.Times.Exceptions;
+
+namespace Tarteeb.Api.Services.Foundations.Times
+{
+    public partial class TimeService
+    {
+        private void ValidateTimeId(Guid timeId) =>
+            Validate((IsInvalid(timeId), nameof(Time.Id)));
```
Comment (GulchexrasGithub) [inline Tarteeb.Api/Services/Foundations/Times/TimeService.Validations.cs:17]: `Validate((Rule: IsInvalid(timeId), Parametr: nameof(Time.Id)));

File: Tarteeb.Api/Services/Foundations/Times/TimeService.Exceptions.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            }
+            catch (NotFoundTimeException notFoundTimeException)
+            {
+                throw CreateAndLogValidationException(notFoundTimeException);
+            }
+            catch (DbUpdateConcurrencyException dbUpdateConcurrencyException)
+            {
+                var lockedTimeException = new LockedTimeException(dbUpdateConcurrencyException);
+
+                throw CreateAndDependencyValidationException(lockedTimeException);
```
Comment (GulchexrasGithub) [inline Tarteeb.Api/Services/Foundations/Times/TimeService.Exceptions.cs:38]: `CreateAndDependencyValidationException` => `CreateAndLogDependencyValidationException`

File: Tarteeb.Api/Services/Foundations/Times/TimeService.Exceptions.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            catch (Exception serviceException)
+            {
+                var failedTimeServiceException = new FailedTimeServiceException(serviceException);
+
+                throw CreateAndLogServiceException(failedTimeServiceException);
+            }
+        }
+
+        private TimeServiceException CreateAndLogServiceException(
+            Exception exception)
```
Comment (GulchexrasGithub) [inline Tarteeb.Api/Services/Foundations/Times/TimeService.Exceptions.cs:56]: Exception emas Xeption

File: Tarteeb.Api.Tests.Unit/Services/Foundations/TimeSlots/TimeServiceTests.Validations.RemoveById.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            // then
+            actualTimeValidationException.Should().BeEquivalentTo(expectedTimeValidationException);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectTimeByIdAsync(inputTimeId), Times.Once);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedTimeValidationException))), Times.Once);
+
```
Comment (GulchexrasGithub) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/TimeSlots/TimeServiceTests.Validations.RemoveById.cs:91]: `this.storageBrokerMock.Verify(broker =>
				broker.DeleteTimeAsync(It.IsAny<Time>()),
					Times.Never);`

File: Tarteeb.Api.Tests.Unit/Services/Foundations/TimeSlots/TimeServiceTests.Exceptions.RemoveById.cs [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowDependencyExceptionOnDeleteWhenSqlExceptionOccursAndLogItAsync()
+        {
+            // given
+            Guid someTimeId = Guid.NewGuid();
+            var sqlException = CreateSqlException();
```
Comment (GulchexrasGithub) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/TimeSlots/TimeServiceTests.Exceptions.RemoveById.cs:69]: SqlException sqlException

File: Tarteeb.Api.Tests.Unit/Services/Foundations/TimeSlots/TimeServiceTests.Exceptions.RemoveById.cs [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+            TimeDependencyException actualTimeDependencyException =
+                await Assert.ThrowsAsync<TimeDependencyException>(
+                    removeTimeByIdTask.AsTask);
+
+            // then
+            actualTimeDependencyException.Should()
+                .BeEquivalentTo(expectedTimeDependencyException);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectTimeByIdAsync(someTimeId), Times.Once);
```
Comment (GulchexrasGithub) [inline Tarteeb.Api.Tests.Unit/Services/Foundations/TimeSlots/TimeServiceTests.Exceptions.RemoveById.cs:94]: `broker.SelectTimeByIdAsync(someTimeId), Times.Once);` => `broker.SelectTimeByIdAsync(It.IsAny<Guid>()), Times.Once);`

**Review by GulchexrasGithub — CHANGES_REQUESTED** [review]
TimeServiceTests classidagi hamma funksyalar static bo'lishi kerak.

**Comment by AbbosHaydarov** [comment]
> TimeServiceTests classidagi hamma funksyalar static bo'lishi kerak.

Resolved All of Comments

File: Tarteeb.Api/Services/Foundations/Times/TimeService.Exceptions.cs [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+        private TimeDependencyException CreateAndLogCriticalDependencyException(Xeption exception)
+        {
+            var timeDependecyException = new TimeDependencyException(exception);
+            this.loggingBroker.LogCritical(timeDependecyException);
+
+            return timeDependecyException;
+        }
+
+        private TimeDependencyValidationException CreateAndDependencyValidationException(Xeption exception)
```
Comment (GulchexrasGithub) [inline Tarteeb.Api/Services/Foundations/Times/TimeService.Exceptions.cs:80]: `CreateAndDependencyValidationException` => `CreateAndLogDependencyValidationException`

Comment (Nodirkhan) [reply Tarteeb.Api/Services/Foundations/Times/TimeService.Exceptions.cs]: Done

**Review by GulchexrasGithub — APPROVED** [review]

**Comment by GulchexrasGithub** [comment]
Again conflicts

**Review by jasurbekyusuf — COMMENTED** [review]
LGTM

File: Tarteeb.Api/Models/Foundations/Times/Exceptions/NotFoundTimeException.cs [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+using System;
+using Xeptions;
+
+namespace Tarteeb.Api.Models.Foundations.Times.Exceptions
+{
+    public class NotFoundTimeException : Xeption
+    {
+        public NotFoundTimeException(Guid timeId)
+            : base(message: $"Couldn't find team with id: {timeId}.")
```
Comment (ElbekDeveloper) [inline Tarteeb.Api/Models/Foundations/Times/Exceptions/NotFoundTimeException.cs:14]: team ? => time bo'ladi

File: Tarteeb.Api/Services/Foundations/Times/TimeService.Exceptions.cs [hunk 12]
```diff
@@ -14,8 +16,44 @@ namespace Tarteeb.Api.Services.Foundations.Times
...
+                throw CreateAndLogDependencyValidationException(lockedTimeException);
+            }
+            catch (SqlException sqlException)
+            {
+                var failedTimeStorageException =
+                    new FailedTimeStorageException(sqlException);
+
+                throw CreateAndLogCriticalDependencyException(failedTimeStorageException);
+            }
+            catch (Exception serviceException)
```
Comment (ElbekDeveloper) [inline Tarteeb.Api/Services/Foundations/Times/TimeService.Exceptions.cs:49]: Birinchi SQL exception handle qilinadi. Keyinhchalik dbupdate exception. Bu yerda va testda ham to'g'rilang

**Review by ElbekDeveloper — APPROVED** [review]

**Pull request merged**